src/loader.rs

```diff
@@ -48,25 +41,34 @@ pub(crate) fn load(
 pub(crate) fn load_bytes(
     page_table: &mut LoaderPageTable,
     bytes: &[u8],
-) -> Result<u64> {
+) -> Result<*const u8> {
     let elf = parse_elf(bytes)?;
-    for section in elf.program_headers.iter().filter(|&h| h.p_type == PT_LOAD) {
-        let file_range = section.file_range();
-        if bytes.len() < file_range.end {
+    load(page_table, &elf, &bytes)
+}
+
+fn load(
+    page_table: &mut LoaderPageTable,
+    elf: &Elf<'_>,
+    file: &dyn Read,
+) -> Result<*const u8> {
+    let mut entry = ptr::null();
+    let elfentry = elf.entry.try_into().unwrap();
+    for segment in elf.program_headers.iter().filter(|&h| h.p_type == PT_LOAD) {
+        let file_range = segment.file_range();
+        if file.size() < file_range.end {
             return Err(Error::ElfTruncatedObj);
         }
-        load_segment(page_table, section, &bytes)?;
+        let (base, len) = load_segment(page_table, segment, file)?;
+        let addr = base.addr();
+        let mem_range = addr..addr + len;
+        if mem_range.contains(&elfentry) {
+            entry = base.with_addr(elfentry);
+        }
```

Comment:
nice. i'd suggested the provenance of the page table only because i wasn't immediately sure what would be better, but this makes much more sense.